### PR TITLE
Fix AccessViolation/InvalidProgramException in custom nullable converters.

### LINKED ISSUE #55135: Json serialize when use custom nullable JsonConverter always thrown AccessViolationException
### Description

When use nullable JsonConverter to custom serialize double? and double, always thrown  AccessViolationException.

### Configuration

|Item|Version|
|:-:|:-:|
|.NET Runtime|5.0.301|
|TargetFramework|5.0|
|System.Text.Json|5.0.1|

### Regression?

Never normal

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Fixes a condition where a custom nullable type converter can result in `AccessViolationException`/`InvalidProgramExceptions` when the user has opted into the underlying struct type via `CanConvert` overrides.

We already have logic for detecting the converse, namely struct converters opting into nullable converters, so this PR is merely extending the existing checks.

Fixes #55135.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitMemberAccessor.cs [hunk 1]
```diff
@@ -258,6 +258,10 @@ private static DynamicMethod CreatePropertyGetter(PropertyInfo propertyInfo, Typ
 
             if (declaredPropertyType != runtimePropertyType && declaredPropertyType.IsValueType)
             {
+                // Not supported scenario: possible if declaredPropertyType == int? and runtimePropertyType == int
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitMemberAccessor.cs:261]: Alternatively, we could decide to support this scenario by doing something like:
```csharp
                if (runtimePropertyType.IsValueType)
                {
                    Debug.Assert(declaredPropertyType.IsNullableOfT() || runtimePropertyType.IsNullableOfT());
                    generator.Emit(OpCodes.Box, declaredPropertyType);
                    generator.Emit(OpCodes.Unbox_Any, runtimePropertyType);
                }
                else
                {
                    generator.Emit(OpCodes.Box, declaredPropertyType);
                }
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs [hunk 2]
```diff
@@ -128,16 +128,18 @@ internal JsonConverter DetermineConverter(Type? parentClassType, Type runtimePro
...
+            // User has indicated that either:
+            //   a) a non-nullable-struct handling converter should handle a nullable struct type or
+            //   b) a nullable-struct handling converter should handle a non-nullable struct type.
+            // User should implement a custom converter for the underlying struct and remove the unnecessary CanConvert method override.
+            // The serializer will automatically wrap the custom converter with NullableConverter<T>.
             //
             // We also throw to avoid passing an invalid argument to setters for nullable struct properties,
             // which would cause an InvalidProgramException when the generated IL is invoked.
-            // This is not an issue of the converter is wrapped in NullableConverter<T>.
-            if (runtimePropertyType.CanBeNull() && !converter.TypeToConvert.CanBeNull())
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:138]: @layomia @steveharter Based on the commends above I _think_ that this check should use `IsNullableOfT()` which is what the original PR adding this check was doing: https://github.com/dotnet/runtime/commit/99829d5017b274537ada263ea7002ec576fa85d5

I think it was changed accidentally to `CanBeNull()` during a refactoring.

**Comment by eiriktsarpalis** [comment]
Tagging @GrabYourPitchforks who expressed interested in this PR.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs [hunk 3]
```diff
@@ -128,16 +128,18 @@ internal JsonConverter DetermineConverter(Type? parentClassType, Type runtimePro
...
+            if (runtimePropertyType.IsValueType && converter.IsValueType &&
+                runtimePropertyType.IsNullableOfT() ^ converter.TypeToConvert.IsNullableOfT())
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:140]: ^ does have higher precedence than &&, but you need a masters degree in C# to know that :)  Mind adding some parens to make it clear?

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs]: Addressed :-)

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitMemberAccessor.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitMemberAccessor.cs:261]: I think it's better to disallow it as done in this PR and have users provide converter implementations that align with what we intend. This way we can keep IL logic as simplified as possible and reduce the possibility of type errors.

- use a `JsonConverter<double>` if you want custom support for `double` and `double?` (supported by the serializer wrapping custom converter in `NullableConverter<>`)
- use a `JsonConverter<double?>` if you want custom support for `double?` specifically (for whatever reason). A separate `JsonConverter<double>` will be needed for `double` support in this case.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs (hunk 2 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:138]: I believe you're right.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitMemberAccessor.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitMemberAccessor.cs:261]: With your fix, do we still have coverage for this branch? Same question for the location of the new debug assert below.

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/CustomConverterTests/CustomConverterTests.BadConverters.cs [hunk 4]
```diff
@@ -419,5 +419,82 @@ public override void Write(Utf8JsonWriter writer, ClassWithConverterWithoutPubli
                 throw new NotImplementedException();
             }
         }
+
+
+        [Fact]
+        public static void BadDoubleConverter_Serialization()
```
Comment (layomia) [inline src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/CustomConverterTests/CustomConverterTests.BadConverters.cs:425]: nit: can we have comments or update the test method and/or test type names to indicate whey the converters are "bad"? (Faulty `CanConvert(Type)` implementations...)

**Review by layomia — APPROVED** [review]
LGTM

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitMemberAccessor.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitMemberAccessor.cs:261]: Yeah, they're both covered by the new tests I added.

**Pull request merged**